File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/base/BaseActivity.java
package org.mifos.mobilewallet.mifospay.base;

import android.app.ProgressDialog;
import android.content.pm.ActivityInfo;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import com.mifos.mobile.passcode.BasePassCodeActivity;

import org.mifos.mobilewallet.mifospay.MifosPayApp;
import org.mifos.mobilewallet.mifospay.R;
import org.mifos.mobilewallet.mifospay.injection.component.ActivityComponent;
import org.mifos.mobilewallet.mifospay.injection.component.DaggerActivityComponent;
import org.mifos.mobilewallet.mifospay.injection.module.ActivityModule;
import org.mifos.mobilewallet.mifospay.passcode.ui.PassCodeActivity;

/**
 * Created by naman on 16/6/17.
 */

public class BaseActivity extends BasePassCodeActivity implements BaseActivityCallback {

    public Toolbar toolbar;
    public SwipeRefreshLayout swipeLayout;
    public ProgressDialog progressDialog;
    private ActivityComponent activityComponent;

    public ActivityComponent getActivityComponent() {
        if (activityComponent == null) {
            activityComponent = DaggerActivityComponent.builder()
                    .activityModule(new ActivityModule(this))
                    .applicationComponent(MifosPayApp.get(this).component())
                    .build();
        }
        return activityComponent;
    }

    @Override
    public void setContentView(int layoutResID) {
        super.setContentView(layoutResID);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        toolbar = findViewById(R.id.toolbar);
        swipeLayout = findViewById(R.id.swipe_layout);
        if (toolbar != null) {
            setSupportActionBar(toolbar);
        }
    }

    @Override
    public void showSwipeProgress() {
        if (swipeLayout != null) {
            swipeLayout.setEnabled(true);
            swipeLayout.setRefreshing(true);
        }
    }

    @Override
    public void hideSwipeProgress() {
        if (swipeLayout != null) {
            swipeLayout.setRefreshing(false);
        }
    }

    @Override
    public void setSwipeRefreshEnabled(boolean enabled) {
        if (swipeLayout != null) {
            swipeLayout.setEnabled(enabled);
        }
    }

    @Override
    public void showProgressDialog(String message) {
        if (progressDialog != null) {
            progressDialog.setMessage(message);
            progressDialog.show();
        } else {
            progressDialog = new ProgressDialog(this);
            progressDialog.setCanceledOnTouchOutside(false);
            progressDialog.setMessage(message);
            progressDialog.show();
        }
    }

    @Override
    public void hideProgressDialog() {
        if (progressDialog != null) {
            progressDialog.hide();
        }
    }

    @Override
    public void cancelProgressDialog() {
        if (progressDialog != null && progressDialog.isShowing()) {
            progressDialog.cancel();
        }
    }
    @Override
    public void dismissProgressDialog() {
        if (progressDialog != null && progressDialog.isShowing()) {
            progressDialog.dismiss();
        }
    }

    @Override
    public void setToolbarTitle(String title) {
        if (getSupportActionBar() != null && getTitle() != null) {
            setTitle(title);
        }
    }

    @Override
    public void showColoredBackButton(int drawable) {
        showHomeButton();
        setToolbarIcon(drawable);
    }

    @Override
    public void showCloseButton() {
        showHomeButton();
        setToolbarIcon(R.drawable.ic_close);
    }

    private void showHomeButton() {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }

    private void setToolbarIcon(int res) {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setHomeAsUpIndicator(res);
        }
    }

    @Override
    public void hideBackButton() {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setHomeButtonEnabled(false);
            getSupportActionBar().setDisplayHomeAsUpEnabled(false);
        }
    }

    @Override
    public SwipeRefreshLayout getSwipeRefreshLayout() {
        return swipeLayout;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    public void addFragment(Fragment fragment, int containerId) {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.add(containerId, fragment);
        transaction.commit();
    }

    public void replaceFragment(Fragment fragment, boolean addToBackStack, int containerId) {
        invalidateOptionsMenu();
        String backStateName = fragment.getClass().getName();
        boolean fragmentPopped = getSupportFragmentManager().popBackStackImmediate(backStateName,
                0);

        if (!fragmentPopped && getSupportFragmentManager().findFragmentByTag(backStateName) ==
                null) {
            FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
            transaction.setCustomAnimations(R.anim.custom_fade_in, R.anim.custom_fade_out);
            transaction.replace(containerId, fragment, backStateName);
            if (addToBackStack) {
                transaction.addToBackStack(backStateName);
            }
            transaction.commit();
        }
    }

    public void clearFragmentBackStack() {
        FragmentManager fm = getSupportFragmentManager();
        int backStackCount = getSupportFragmentManager().getBackStackEntryCount();
        for (int i = 0; i < backStackCount; i++) {
            int backStackId = getSupportFragmentManager().getBackStackEntryAt(i).getId();
            fm.popBackStack(backStackId, FragmentManager.POP_BACK_STACK_INCLUSIVE);
        }
    }

    @Override
    public Class getPassCodeClass() {
        return PassCodeActivity.class;
    }
}

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/base/BaseActivity.kt
package org.mifos.mobilewallet.mifospay.base

import android.app.ProgressDialog
import android.content.pm.ActivityInfo
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.Toolbar
import android.view.MenuItem
import com.mifos.mobile.passcode.BasePassCodeActivity
import org.mifos.mobilewallet.mifospay.MifosPayApp
import org.mifos.mobilewallet.mifospay.R
import org.mifos.mobilewallet.mifospay.injection.component.ActivityComponent
import org.mifos.mobilewallet.mifospay.injection.component.DaggerActivityComponent
import org.mifos.mobilewallet.mifospay.injection.module.ActivityModule
import org.mifos.mobilewallet.mifospay.passcode.ui.PassCodeActivity

/**
 * Created by naman on 16/6/17.
 */
open class BaseActivity : BasePassCodeActivity(), BaseActivityCallback {
    var toolbar: Toolbar? = null
    override var swipeRefreshLayout: SwipeRefreshLayout? = null
    var progressDialog: ProgressDialog? = null
    var activityComponent: ActivityComponent? = null
        get() {
            if (field == null) {
                field = DaggerActivityComponent.builder()
                    .activityModule(ActivityModule(this))
                    .applicationComponent(MifosPayApp.get(this).component())
                    .build()
            }
            return field
        }
        private set

    override fun setContentView(layoutResID: Int) {
        super.setContentView(layoutResID)
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
        toolbar = findViewById(R.id.toolbar)
        swipeRefreshLayout = findViewById(R.id.swipe_layout)
        if (toolbar != null) {
            setSupportActionBar(toolbar)
        }
    }

    override fun showSwipeProgress() {
        if (swipeRefreshLayout != null) {
            swipeRefreshLayout?.isEnabled = true
            swipeRefreshLayout?.isRefreshing = true
        }
    }

    override fun hideSwipeProgress() {
        if (swipeRefreshLayout != null) {
            swipeRefreshLayout?.isRefreshing = false
        }
    }

    override fun setSwipeRefreshEnabled(enabled: Boolean) {
        if (swipeRefreshLayout != null) {
            swipeRefreshLayout?.isEnabled = enabled
        }
    }

    override fun showProgressDialog(message: String?) {
        if (progressDialog != null) {
            progressDialog?.setMessage(message)
            progressDialog?.show()
        } else {
            progressDialog = ProgressDialog(this)
            progressDialog?.setCanceledOnTouchOutside(false)
            progressDialog?.setMessage(message)
            progressDialog?.show()
        }
    }

    override fun hideProgressDialog() {
        if (progressDialog != null) {
            progressDialog?.hide()
        }
    }

    override fun cancelProgressDialog() {
        if (progressDialog != null && progressDialog?.isShowing == true) {
            progressDialog?.cancel()
        }
    }

    override fun dismissProgressDialog() {
        if (progressDialog != null && progressDialog?.isShowing == true) {
            progressDialog?.dismiss()
        }
    }

    override fun setToolbarTitle(title: String?) {
        if (supportActionBar != null && getTitle() != null) {
            setTitle(title)
        }
    }

    override fun showColoredBackButton(drawable: Int) {
        showHomeButton()
        setToolbarIcon(drawable)
    }

    override fun showCloseButton() {
        showHomeButton()
        setToolbarIcon(R.drawable.ic_close)
    }

    private fun showHomeButton() {
        if (supportActionBar != null) {
            supportActionBar?.setHomeButtonEnabled(true)
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }
    }

    private fun setToolbarIcon(res: Int) {
        if (supportActionBar != null) {
            supportActionBar?.setHomeAsUpIndicator(res)
        }
    }

    override fun hideBackButton() {
        if (supportActionBar != null) {
            supportActionBar?.setHomeButtonEnabled(false)
            supportActionBar?.setDisplayHomeAsUpEnabled(false)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }

    fun addFragment(fragment: Fragment?, containerId: Int) {
        val transaction = supportFragmentManager.beginTransaction()
        fragment?.let { transaction.add(containerId, it) }
        transaction.commit()
    }

    fun replaceFragment(fragment: Fragment, addToBackStack: Boolean, containerId: Int) {
        invalidateOptionsMenu()
        val backStateName = fragment.javaClass.name
        val fragmentPopped = supportFragmentManager.popBackStackImmediate(
            backStateName,
            0
        )
        if (!fragmentPopped && supportFragmentManager.findFragmentByTag(backStateName) ==
            null
        ) {
            val transaction = supportFragmentManager.beginTransaction()
            transaction.setCustomAnimations(R.anim.custom_fade_in, R.anim.custom_fade_out)
            transaction.replace(containerId, fragment, backStateName)
            if (addToBackStack) {
                transaction.addToBackStack(backStateName)
            }
            transaction.commit()
        }
    }

    fun clearFragmentBackStack() {
        val fm = supportFragmentManager
        val backStackCount = supportFragmentManager.backStackEntryCount
        for (i in 0 until backStackCount) {
            val backStackId = supportFragmentManager.getBackStackEntryAt(i).id
            fm.popBackStack(backStackId, FragmentManager.POP_BACK_STACK_INCLUSIVE)
        }
    }

    override fun getPassCodeClass(): Class<*> {
        return PassCodeActivity::class.java
    }
}